include/chain_invoke.h

```diff
@@ -61,18 +61,38 @@ using namespace msl;
             }
         }
 
+        template<bool have_agrs, typename TRet, typename stack_args>
+        struct TypeSelector final
+        {
+            private:
+                using t1 = std::conditional_t<have_agrs, stack_args, std::tuple<TRet>>;
+                inline static constexpr bool is_ref = std::is_reference_v<TRet>;
+                using no_ref_but_pointer = std::tuple<std::add_pointer_t<std::remove_reference_t<TRet>>>;
+                using t2 = std::conditional_t<is_ref, std::tuple<no_ref_but_pointer>, std::tuple<TRet>>;
+
+            public:
+                inline static constexpr bool have_agrs_value = have_agrs;
+                inline static constexpr bool is_ref_value = is_ref;
+                using type = std::conditional_t<is_ref, t2, t1>;
```

Comment:
You're having the same condition `is_ref `here and line 71 (`using t2...`), so `t2` is always `std::tuple<no_ref_but_pointer>`

Reply:
I did changes and comments. I don't know how do better. You can see after i upload changes

Reply:
I got the intension after reading the comment above.
And regarding 

> I don't know how do better. 

Try [this](https://godbolt.org/z/hMhco5) new constexpr metaprogramming. Maybe it will be better.

Reply:
new constexpr metaprogramming is good. But we need not only type select. Also we need know conditions  below.
For example OwningInvokingStep work in diffirent compile time branches. (stakc_args, ret value, ret pointer, ret reference)
I will think more at this place, but probably can't do a solution you will like